.github/workflows/integration.yml

```diff
@@ -40,94 +43,342 @@ jobs:
-          - test_name: 'prometheus'
-            if: ${{ github.event_name == 'merge_group' || needs.changes.outputs.int-all == 'true' || needs.changes.outputs.prometheus == 'true' }}
-          - test_name: 'pulsar'
-            if: ${{ github.event_name == 'merge_group' || needs.changes.outputs.int-all == 'true' || needs.changes.outputs.pulsar == 'true' }}
-          - test_name: 'redis'
-            if: ${{ github.event_name == 'merge_group' || needs.changes.outputs.int-all == 'true' || needs.changes.outputs.redis == 'true' }}
-          - test_name: 'shutdown'
-            if: ${{ github.event_name == 'merge_group' || needs.changes.outputs.int-all == 'true' }}
-          - test_name: 'splunk'
-            if: ${{ github.event_name == 'merge_group' || needs.changes.outputs.int-all == 'true' || needs.changes.outputs.splunk == 'true' }}
-          - test_name: 'webhdfs'
-            if: ${{ github.event_name == 'merge_group' || needs.changes.outputs.int-all == 'true' || needs.changes.outputs.webhdfs == 'true' }}
-
-  # This is a required status check, so it always needs to run if prior jobs failed, in order to mark the status correctly.
-  integration:
+    if: |
+      github.event_name == 'merge_group' || (
+        needs.changes.outputs.int-all == 'true'
+          || needs.changes.outputs.amqp == 'true'
+          || needs.changes.outputs.appsignal == 'true'
+          || needs.changes.outputs.aws == 'true'
+          || needs.changes.outputs.axiom == 'true'
+          || needs.changes.outputs.azure == 'true'
+          || needs.changes.outputs.clickhouse == 'true'
+          || needs.changes.outputs.databend == 'true'
+          || needs.changes.outputs.datadog == 'true'
+          || needs.changes.outputs.dnstap == 'true'
+          || needs.changes.outputs.docker-logs == 'true'
+          || needs.changes.outputs.elasticsearch == 'true'
+          || needs.changes.outputs.eventstoredb == 'true'
+          || needs.changes.outputs.fluent == 'true'
+          || needs.changes.outputs.gcp == 'true'
+          || needs.changes.outputs.humio == 'true'
+          || needs.changes.outputs.http-client == 'true'
+          || needs.changes.outputs.influxdb == 'true'
+          || needs.changes.outputs.kafka == 'true'
+          || needs.changes.outputs.logstash == 'true'
+          || needs.changes.outputs.loki == 'true'
+          || needs.changes.outputs.mongodb == 'true'
+          || needs.changes.outputs.nats == 'true'
+          || needs.changes.outputs.nginx == 'true'
+          || needs.changes.outputs.opentelemetry == 'true'
+          || needs.changes.outputs.postgres == 'true'
+          || needs.changes.outputs.prometheus == 'true'
+          || needs.changes.outputs.pulsar == 'true'
+          || needs.changes.outputs.redis == 'true'
+          || needs.changes.outputs.splunk == 'true'
+          || needs.changes.outputs.webhdfs == 'true'
+      )
+    timeout-minutes: 60
```

Comment:
The tradeoff is known yes. In general, from the CI cost reduction effort we thus far have been accepting some increased runtime at the <pun> expense , of runtime as long as it isn't highly impacting DX.

> (1) to break this up into multiple workflows (each workflow would run n integration tests) when the run time becomes unacceptable 

This would be pretty easy- actually we could do that in the same workflow, just have two jobs and not have them depend on eachother. The two would start in parallel.

> (2) improve caching / environment sharing to minimize environment setup time, which could allow us to run these tests in parallel again.

It did cross my mind that we could potentially re-introduce the matrix approach if we solve the caching issue. The main caveat to that here is that the int tests are all run in containers so we'd have to essentially cache a container image, which would be a bit different than just caching build artifacts.

These are good thoughts. Option 1 is a "cheap and easy" solution to that.

Reply:
More thoughts coming to me- if we made two jobs in the same workflow, they would both take the time to build the image, so yeah that would only benefit once we crossed the idk like 60-70 min total runtime (essentially when we have enough int tests that the running of just all of those together exceeds the time to build).

A side note, @jszwedko and I also discussed the option of moving the int tests back to running on all PR pushes, if we had good cost reduction on it. (or at least on changes to src/lib). But imo I kind of like the on-file detection for PR pushes personally.

Reply:
> Re (2) above I think the container reuses the source directory to do the cargo build, which includes target.

😮 If that is the case, then yes, that would be less of a complexity/ difference from the other cache needs we have than I was thinking.

> would only need to cache the CI step

You lost me there... 🤔  what is the "CI step" you are referring to?

Reply:
To test, run `cargo build --features …` and then `vdev int test …` and see how much it rebuilds.

By "CI step" I was thinking we could do the build in one step, and then run the actual integration tests in parallel running out of that pre-built binary.

Reply:
👍 got it
Will remember this when spiking/undertaking the caching effort.

For the changes in this PR, I think we are good. This discussion stemmed from a (valid) theoretic / future consideration where cross a threshold of having enough integrations that the runtime of doing all those post-build exceeds the time to do a second build / runtime pains.

But, if we get caching going, it would def be worth the runtime improvements to run these in a matrix with the cached binary, in addition to reducing the length of the workflow file(s).